src/librustc/ty/sty.rs

```diff
@@ -927,11 +888,15 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
         }
     }
 
-    pub fn is_empty(&self, _cx: TyCtxt) -> bool {
+    pub fn is_empty(&self, cx: TyCtxt) -> bool {
         // FIXME(#24885): be smarter here
         match self.sty {
             TyEnum(def, _) | TyStruct(def, _) => def.is_empty(),
-            _ => false
+            TyEmpty => true,
+            TyTuple(ref tys) => tys.iter().any(|ty| ty.is_empty(cx)),
```

Comment:
> Why did you make this recursive?

To make it smarter, as the FIXME suggests. And because some tuples are empty.

> Also, you *shouldn't* use the same method for checking for `enum Void {}` and `!`. It would mean `Void` gets the same semantics as `!` which is not what you want.

The are some cases where we do want `enum Void {}` and `!` to have the same semantics. For example, a function that returns *any* empty type should be marked with the `NoReturn` attribute.

> My suggestion is to use `is_never` for `!` and leave `is_empty` untouched.

Well there's other places where I check for `!` specifically by matching on `sty`.